File: android/app/src/main/java/com/yoginiastro/incoming/IncomingChatDisplay.kt
package com.yoginiastro.incoming

object IncomingChatDisplay {

  fun resolveTitle(payload: Map<String, String>): String {
    return payload["title"]?.trim()?.takeIf { it.isNotEmpty() }
      ?: payload["customerName"]?.trim()?.takeIf { it.isNotEmpty() }
      ?: payload["senderName"]?.trim()?.takeIf { it.isNotEmpty() }
      ?: "Incoming chat request"
  }

  fun resolveBody(payload: Map<String, String>): String {
    payload["body"]?.trim()?.takeIf { it.isNotEmpty() }?.let { return it }
    payload["message"]?.trim()?.takeIf { it.isNotEmpty() }?.let { return it }
    payload["subtitle"]?.trim()?.takeIf { it.isNotEmpty() }?.let { return it }
    payload["notificationBody"]?.trim()?.takeIf { it.isNotEmpty() }?.let { return it }
    val waiting = payload["waitingCount"]?.trim()?.takeIf { it.isNotEmpty() }
    if (waiting != null) {
      val n = waiting.toIntOrNull()
      if (n != null && n > 0) {
        return "Users waiting: $n"
      }
    }
    return "Wants to chat with you"
  }
}

File: src/services/push/incomingChatDisplay.ts
import type { OrderStackParamList } from '../../navigation/types';

type DisplayFields = Pick<
  OrderStackParamList['IncomingChatRequest'],
  'customerName' | 'notificationTitle' | 'notificationBody' | 'message' | 'subtitle'
> & {
  waitingCount?: string | number;
};

/** Notification / overlay title — always prefers live FCM user name. */
export function resolveIncomingChatTitle(input: DisplayFields): string {
  return (
    input.notificationTitle?.trim() ||
    input.customerName?.trim() ||
    'Incoming chat request'
  );
}

/** Notification / overlay subtitle line from FCM message, waitlist count, etc. */
export function resolveIncomingChatBody(input: DisplayFields): string {
  const explicit =
    input.notificationBody?.trim() ||
    input.message?.trim() ||
    input.subtitle?.trim();
  if (explicit) {
    return explicit;
  }
  const waiting = input.waitingCount;
  if (waiting !== undefined && waiting !== '') {
    const n = Number(waiting);
    if (!Number.isNaN(n) && n > 0) {
      return `Users waiting: ${n}`;
    }
  }
  return 'Wants to chat with you';
}
